### PR TITLE
[Feature] Radix Tree in C++

### PR DESCRIPTION
**Pull request by DarkSharpness**
## Motivation

This PR is still part of #7194. 

## Modifications

- **Upstream Hierarchical Radix Tree in C++:**  
  Added the core radix tree implementation for future hierarchical caching.
- **Python Interface (Non-Hierarchical Only):**  
  Implemented Python bindings for the radix tree, currently supporting only non-hierarchical caching. Hierarchical cache integration will be addressed in subsequent PRs.
- **Experimental Feature Flag:**  
  The experimental C++ radix tree can be enabled by setting the environment variable `SGLANG_EXPERIMENTAL_HIRADIX_CACHE=1`. Default behavior remains unchanged.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by KevinZeng08** [comment]
Great work! 
I would like to ask how much the performance of the C++ version of RadixTree can be expected to improve?

**Comment by DarkSharpness** [comment]
> Great work! I would like to ask how much the performance of the C++ version of RadixTree can be expected to improve?

Thank you for your question! Based on my testing, the C++ implementation typically delivers **2-3× speedups** over the Python version. In extreme cases like #6907, this should significantly reduce `evict` operation overhead.

Note that:
1. This PR is part of #7194, which is not meant to replace the original radix cache in Python. Its ultimate goal is to accelerate some advanced prefix-matching related scheduling when hierarchical cache is enabled.
2. The C++ implementation is actually 10 times faster than Python Radix Cache when tested alone. But in production, some other operations like `torch.cat` will amortize the speedup.
3. Current Python implementation works fine for most workloads.

File: python/sglang/srt/managers/scheduler.py [hunk 1]
```diff
@@ -577,7 +578,20 @@ def init_memory_pool_and_cache(self):
                 page_size=self.page_size,
             )
         else:
-            if self.enable_hierarchical_cache:
+            if os.environ.get("SGLANG_EXPERIMENTAL_HIRADIX_CACHE") == "1":
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]:
```suggestion
            if os.environ.get("SGLANG_EXPERIMENTAL_CPP_RADIX_TREE") == "1":
```

File: python/sglang/srt/managers/scheduler.py [hunk 2]
```diff
@@ -577,7 +578,20 @@ def init_memory_pool_and_cache(self):
...
+                self.tree_cache = HiRadixCacheCpp(
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]:
```suggestion
                self.tree_cache = RadixCacheCpp(
```

**Review by merrymercy — CHANGES_REQUESTED** [review]
1. move this to JIT https://docs.pytorch.org/tutorials/advanced/cpp_extension.html#jit-compiling-extensions. Only jit/import when it is really used. Open a new folder `python/sglang/srt/mem_cache/cpp_radix_tree`
2. Add an e2e test case `test/srt/test_cpp_radix_tree.py`. Follow this https://github.com/sgl-project/sglang/blob/83646089301a83b55845e48e93be0cefb6a21a7b/test/srt/test_page_size.py#L16
    - Test e2e accuracy
    - Test cache hit rate.

**Comment by DarkSharpness** [comment]
All fixed. cc @merrymercy @xiezhq-hermann

**Review by merrymercy — APPROVED** [review]
